schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,292 @@
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
+  bitWidth Int
+  bucketSize Int
+  map Bytes
+  data [ Element ]
+}
+
+# Non-root node layout
+type HashMapNode struct {
+  map Bytes
+  data [ Element ]
+}
+
+type Element union {
+  | Link "link"
+  | Bucket "bucket"
+} representation keyed
+
+type Bucket list [ BucketEntry ]
+
+type BucketEntry struct {
+  key Bytes
+  value Value
+} representation tuple
+
+type Value union {
+  | Bool bool
+  | String string
+  | Bytes bytes
+  | Int int
+  | Float float
+  | Map map
+  | List list
+  | Link link
+} representation kinded
+```
+
+Notes:
+
+* `hashAlg` in the root block is a string name that corresponds to a [multihash](https://github.com/multiformats/multihash) as found in the [multiformats table](https://github.com/multiformats/multicodec/blob/master/table.csv).
+* `bitWidth` in the root block must be at least `2` and no more than `12`. ***(TODO: this is arbitrary and currently IAMap only supports a maximum of `8` and go-hamt-ipld _only_ supports `8`. `12` yields a maximum `data` array length of `4096`, is it reasonable to go higher? A two-byte maximum would be neat, but `65,536` elements is maybe too much.)***
+* `bucketSize` in the root block must be at least `1`. ***(TODO: does it need a maximum?)***
+* Values can be any kind but cannot be `Null`
+* Keys are stored in `Byte` form
+
+## Algorithm in detail
+
+### `Get(key)`
+
+1. Set a `depth` value to `0`, indicating the root block
+2. The `key` is hashed, using `hashAlg`.
+3. Take the left-most `bitWidth` bits, offset by `depth x bitWidth`, from the hash to form an `index`. At each level of the data structure, we increment the section of bits we take from the hash so that the `index` comprises a different set of bits as we move down.
+4. If the `index` bit in the node's `map` is `0`, we can be certain that the `key` does not exist in this data structure, so return a `Null` value.
+5. If the `index` bit in the node's `map` is `1`, the value may exist. Perform a `popcount()` on the `map` up to `index` such that we count the number of `1` bits up to the `index` bit-position. This gives us `dataIndex`, an index in the `data` array to look up the value or insert a new bucket.
+6. If the `dataIndex` element of `data` contains a link (CID) to a child block, increment `depth` and repeat with the child node identified by the link from step **3**.
```

Comment:
(keyed union)

Reply:
I have two variants scattered in this section: "contains a link (CID)" and "contains a bucket (array)", assuming that the reader will identify that the discriminator is in the keyed union. Should it be something like "contains a 'link' property" and "contains a 'bucket' property"?